docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,119 @@
+
+## Decision
+
+The Cosmos SDK will define an `AnteHandler` for IBC packet receiving. The `AnteHandler` will iterate over the messages included in the 
+transaction, type switch to check whether the message contains an incoming IBC packet, and if so verify the Merkle proof.
+
+```go
+// Pseudocode
+func IBCAnteHandler(ctx sdk.Context, tx sdk.Tx, bool) (sdk.Context, sdk.Result, bool) {
+  for _, msg := range tx.GetMsgs() {
+    if msg, ok := msg.(MsgPacket); ok {
+      if err := VerifyPacket(msg.Packet, msg.Proofs, msg.ProofHeight); err != nil {
+         return ctx, err.Result(), true
+      }
+    }
+  }
+  return ctx, sdk.Result{}, false
+}
+```
+
+where `MsgPacket` is the `sdk.Msg` type including any IBC packet inside and embedding `Packet.Route()` method.
+
+The `AnteHandler` will be inserted to the top level application, after the signature authentication logic provided by `auth.NewAnteHandler`, utilizing `AnteDecorator` pattern.
+
+The Cosmos SDK will define the wrapper function `ReceivePacket` under the port keeper. The function will wrap packet handlers to automatically handle the acknowledgements.
+
+```go
+// Pseudocode
+func (k PortKeeper) ReceivePacket(ctx sdk.Context, msg MsgPacket, h func(sdk.Context, Packet), sdk.Result) sdk.Result {
+  // Cache context
+  cctx, write := ctx.CacheContext()
+
+  // verification already done inside the antehandler
+  res := h(cctx, msg.Packet)
+  
+  // write the cache only if succedded
+  if res.IsOK() {
+    write()
+  }
+  
+  // set the result to OK to persist the state change
+  res.Code = sdk.CodeOK
+  
+  // res.Data will be stored as acknowledgement; []byte{} will be stored if not exists
+  if res.Data == nil {
+    res.Data = []byte{}
+  }
+  k.SetPacketAcknowledgement(ctx, msg.PortID, msg.ChannelID, msg.Sequence, res.Data)
+
+  return res
+}
+```
+
+Example application-side usage:
+
+```go
+func NewHandler(k Keeper) sdk.Handler {
+  return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
+    switch msg := msg.(type) {
+    case ibc.MsgPacket:
+      return k.port.ReceivePacket(ctx, msg, func(ctx sdk.Context, p Packet) sdk.Result {
+        switch packet := packet.(type) {
+        case MyPacket:
+          return handleMyPacket(ctx, k, packet)
+        }
+      })
+    }
+  }
+}
+
+func handleMyPacket(ctx sdk.Context, k keeper, packet MyPacket) sdk.Result {
+  if failureCondition {
+    return AckInvalidPacketContent(k.codespace, []byte{packet.Data})
+  }
+  return sdk.Result{}
+}
+```
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+- Intuitive interface for developers - IBC handlers do not need to care about IBC authentication
+- State change commitment logic is embedded into `baseapp.runTx` logic
+- IBC applications do not need to define message type for receiving packet
+
+### Negative
+
+- AnteHandler processes over all the transactions before each handler is executed. `UpdateClient` and `RecvPacket` cannot be processed in the same transaction
```

Comment:
I missed this before but this is also quite unfortunate. Is there any way around this that preserves the ergonomics?

Reply:
I think there should be some challenge period between the header submission and the packet receiving for equivocation for the most case, and having two messages separately processed in the chain is better for preventing race condition.

Reply:
Better for preventing what race condition? The proposer will always control the order.

I think there are definitely cases where we want this - consider a solo machine signing with a private key, there is no equivocation challenge period.